internal/pkg/skuba/kubernetes/nodes_test.go

```diff
@@ -18,37 +18,409 @@
+				if actualErr.Error() != expectErr {
+					t.Errorf("returned error (%v) does not match the expected one (%v)", actualErr, expectErr)
+					return
+				}
+				return
+			}
+			if actual.Name != tt.expect {
+				t.Errorf("returned node name (%s) does not match the expected on (%s)", actual.Name, tt.expect)
+				return
+			}
 		})
+	}
+}
 
-		etcdContainerImageTag, _ := getPodContainerImageTag(clientset, namespace, podName)
-		if !reflect.DeepEqual(etcdContainerImageTag, expectedEtcdContainerImageTag) {
-			t.Errorf("got %v, want %v", etcdContainerImageTag, expectedEtcdContainerImageTag)
-		}
-	})
+func Test_DrainNode(t *testing.T) {
+	tests := []struct {
+		name         string
+		node         string
+		reactorPod   string
+		evictSupport bool
+		expectErrMsg string
+	}{
+		{
+			name:         "drain master node by eviction",
+			node:         m1.Name,
+			evictSupport: true,
+			reactorPod:   "valid",
+		},
+		{
+			name:         "drain worker node by eviction",
+			node:         w1.Name,
+			reactorPod:   "valid",
+			evictSupport: true,
+		},
+		{
+			name:       "drain worker node by deletion",
+			node:       w1.Name,
+			reactorPod: "valid",
+		},
+		{
+			name:         "drain when node does not exist",
+			node:         invalid.Name,
+			evictSupport: true,
+			expectErrMsg: fmt.Sprintf("failed to update node status: nodes \"%s\" not found", invalid.Name),
+		},
+		{
+			name:         "drain by eviction when pod does not exist",
+			node:         m1.Name,
+			evictSupport: true,
+			reactorPod:   "invalid",
+			expectErrMsg: "failed to evict pod: invalid: pods \"invalid\" not found",
+		},
+		{
+			name:         "drain by deletion when pod does not exist",
+			node:         m1.Name,
+			reactorPod:   "invalid",
+			expectErrMsg: "failed to delete pod: invalid: pods \"invalid\" not found",
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt // Parallel testing
+		t.Run(tt.name, func(t *testing.T) {
+			fakeClientset := fake.NewSimpleClientset(
+				&corev1.NodeList{
+					Items: []corev1.Node{
+						m1,
+						w1,
+					},
+				},
+				&corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "valid",
+						Namespace: metav1.NamespaceDefault,
+					},
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "test",
+								Image: "registry.suse.com/caasp/v4/test:1.1.1",
+							},
+						},
+					},
+				},
+			)
+			fakeClientset.PrependReactor("list", "pods", func(action ktest.Action) (bool, runtime.Object, error) {
+				obj := &corev1.PodList{
+					Items: []corev1.Pod{
+						{
+							ObjectMeta: metav1.ObjectMeta{
+								Name:      tt.reactorPod,
+								Namespace: metav1.NamespaceDefault,
+							},
+							Spec: corev1.PodSpec{
+								Containers: []corev1.Container{
+									{
+										Name:  "test",
+										Image: "registry.suse.com/caasp/v4/test:1.1.1",
+									},
+								},
+							},
+						},
+					},
+				}
+				return true, obj, nil
+			})
```

Comment:
To be honest, I don't fully understand what this purpose for. But I will not block this PR.

Reply:
The kubernetes fake clientset doesn't handle *selector as expected when called constraints it does not support. 
Instead of returning pods on specific node or errors, it will return all pods as there are no constraints. This leads to a false result when testing case such `drain by eviction when pod does not exist` where expecting errors. 

The address to issue, [here](https://github.com/kubernetes/client-go/issues/326) suggested to use reactor where can implement the expect matching behavior.